chore(accordion): Add React implementation

jpveooys opened the pull request:
This adds a React implementation of the accordion component, using the original GOV.UK JavaScript.

jpveooys commented on src/client/govuk-frontend.d.ts:
```diff
@@ -1,7 +1,13 @@
 declare module 'govuk-frontend' {
-  export class Header<E extends HTMLElement = HTMLElement> {
-    constructor(el: E)
+  class Accordion {
```
Simplified this as the generic didn't seem to be being used.

jpveooys commented on src/nunjucks/moduk/components/accordion/__examples__/with-custom-text.njk:
```diff
@@ -0,0 +1,45 @@
+{% from "moduk/components/accordion/macro.njk" import modukAccordion -%}
```
Added additional Nunjucks examples to mirror the React examples.

jpveooys commented on src/react/accordion/Accordion.tsx:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  return (
+    <AccordionContext.Provider value={{ headingTag, id }}>
```
Use a context here to make these available to accordion items.

jpveooys commented on src/react/accordion/Accordion.tsx:
```diff
@@ -0,0 +1,75 @@
+    <AccordionContext.Provider value={{ headingTag, id }}>
+      <div
+        ref={mergeRefs([ref, forwardedRef])}
+        key={Children.count(children)}
```
As per the original spike – force the component to remount (so the JS reruns)  if the number of items changes.

jpveooys reviewed (commented):
Cancelling empty review (GitHub bug)

jpveooys commented on src/react/accordion/Accordion.tsx:
```diff
@@ -0,0 +1,75 @@
+  hideSectionText?: string
+  hideSectionAriaLabelText?: string
+  id: string
+  rememberExpanded?: boolean
```
Effectively defaults to `true` like the Nunjucks macro (annoying to have a boolean prop that defaults to true though).

jpveooys commented on e2e/tests/accessibility.spec.ts:
```diff
@@ -5,8 +5,7 @@ import { expect, test } from '../fixtures'
   accordion: {
-    default: ['aria-allowed-attr'],
-    'with-summary-lines': ['aria-allowed-attr'],
+    '*': ['aria-allowed-attr'],
```
It was getting a bit annoying having it fail every time I added an example – so added support for `*` meaning all examples for the component.

jpveooys commented on src/react/hooks/useMODUKComponent/useMODUKComponent.ts:
```diff
@@ -1,29 +1,27 @@
-type HeaderConfig = {
-  name: 'Header'
-  element: HTMLElement
+type ComponentElement = {
```
Simplified the types here (so a new type isn't needed for each component).

jpveooys commented on src/client/govuk-frontend.d.ts:
```diff
@@ -1,7 +1,13 @@
-  export class Header<E extends HTMLElement = HTMLElement> {
-    constructor(el: E)
+  class Accordion {
+    constructor(el: HTMLElement)
```
Tried a few different ways to type `el` as `HTMLDivElement` but didn't find something that worked...

jpveooys commented on src/react/accordion/Accordion.tsx:
```diff
@@ -0,0 +1,75 @@
+
+export interface AccordionProps {
+  children: ReactElement<AccordionItemProps>[] | ReactElement<AccordionItemProps>
+  headingTag?: AccordionHeadingTag
```
This prop is slightly different from the Nunjucks macro – it's `headingLevel` there (and a number). Could make it the same, just seemed like extra hassle having them as numbers.

adamledwards commented:
Looks very solid and works really well, 

One issue I spotted is when you replace the content with the same number of children.
The heading does not update.
I did manage to get it working with the snippet below.
To be honest I can't think of a performant way to get round it.
I did consider trying this 
```jsx
const key = useMemo(uniqueId, [children])
<AccordionContext.Provider value={{ headingTag, id }}>
      <div
        key={key}
```
but a re-render higher up the tree will cause `useMODUKComponent` to trigger the js init script
![image](https://github.com/defencedigital/moduk-frontend/assets/5575331/f8dd9529-1c85-46ba-959d-e53b2dfab55e)

jpveooys commented:
@adamledwards Good spot! Also noticed that the type of `children` is a bit inflexible. Will have a look at what can be done without rewriting the JS (though that could still be an option...)

jpveooys commented:
@adamledwards Pushed some updates:

- fixed the children type. Managed to make it accept fragments and still restrict the type of the fragment's children
- made it inspect the props to work out whether to remount. Not so nice, but should keep the scope of the remounts tight. Required using https://github.com/grrowl/react-keyed-flatten-children to flatten the children fully
- fixed a problem with React Testing Library not cleaning up after each test

adamledwards commented on src/react/accordion/Accordion.tsx:
```diff
@@ -24,52 +29,81 @@ export interface AccordionProps {
+    const previousRemountProps = usePrevious(remountProps)
+
+    if (previousRemountProps !== remountProps) {
+      key.current += 1
```
Would this ever be equal? Pick create a new object so I think we would have to use something like https://www.npmjs.com/package/shallow-equal.

jpveooys replied:
Thanks, I was using lodash `isEqual` but accidentally removed it :disappointed: :facepalm: 

Added it back in and added a test for that scenario.

adamledwards reviewed (approved):
👍
